include/circt/Dialect/ESI/ESIServices.h

```diff
@@ -9,14 +9,52 @@
 #ifndef CIRCT_DIALECT_ESI_ESISERVICES_H
 #define CIRCT_DIALECT_ESI_ESISERVICES_H
 
+#include "circt/Dialect/ESI/ESIOps.h"
 #include "circt/Support/LLVM.h"
+
 #include "mlir/Pass/Pass.h"
 #include <memory>
 
 namespace circt {
 namespace esi {
+
+/// Class which "dispatches" a service implementation request to its specified
+/// generator. Also serves as a registery for generators.
+class ServiceGeneratorDispatcher {
+public:
+  // All generators must support this function pointer signature.
+  using ServiceGeneratorFunc =
+      llvm::function_ref<LogicalResult(ServiceImplementReqOp)>;
+
+  // Since passes don't have access to a context at creation time (and
+  // Attributes are tied to the context), we need to delay lookup table creation
+  // until running the dispatch the first time. This is the function pointer
+  // signature to create and return that lookup table.
+  using CreateLookupTable =
+      llvm::function_ref<DenseMap<Attribute, ServiceGeneratorFunc>(
```

Comment:
Instead of mapping from arbitrary Attributes (I see a string attribute called "cosim" used in the example/default), could this be more constrained? Like define an EnumAttr or something where the thing we dispatch on is a sum type.

Reply:
I wanted to provide as much flexibility as possible to the user. (Users will be able to register/provide their own implementation generators.) Registering an Attribute (instead of a StringAttr) gives that flexibility. If user code in a frontend (e.g. PyCDE) wants to register a generator and namespace it, it can register it with a DictAttr like `{ frontend = "pycde", py_callback="gen_telemetry" }`.

Reply:
Ok right that makes sense. For open coding this, using an Attribute seems like the way to go.